src/cosl/coordinated_workers/worker.py

```diff
@@ -501,7 +508,19 @@ def _update_cluster_relation(self) -> None:
         self.cluster.publish_unit_address(socket.getfqdn())
         if self._charm.unit.is_leader() and self.roles:
             logger.info(f"publishing roles: {self.roles}")
-            self.cluster.publish_app_roles(self.roles)
+            try:
+                self.cluster.publish_app_roles(self.roles)
+            except ModelError as e:
+                # if we are handling an event prior to 'start', we could be denied write access
+                # hopefully we can ignore it and trust that it will be retried at the next occasion
```

Comment:
If you really want to do this here, then I'd rather compare envvar against HOOK_INSTALL or something.

In any case, I still think it's better to let the caller guard. When we make such antipattern decisions deep down, production ends up breaking in weird ways, calling for re-relate, refresh "fixes".

Reply:
Thing is there is no hook we can compare against, since the behaviour is "error if we haven't seen install **yet**" and we cannot look into the future. The only solution would be store a history of the events we've seen so far, but that feels un-opsy. So this is the only stateless solution I see.

As for the 'caller should guard' principle, I'm on board, but Worker _is_ the caller.